src/basic/process-util.c

```diff
@@ -158,51 +165,103 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
+                         * and we prefer to get some data rather than none. */
+                        if (ferror(f))
+                                return -errno;
 
-                mfree(t);
-                t = strjoin("[", t2, "]");
-                if (!t)
+                        if (!(flags & PROCESS_CMDLINE_COMM_FALLBACK))
+                                return -ENOENT;
+
+                        /* Kernel threads have no argv[] */
+                        _cleanup_free_ char *t2 = NULL;
+
+                        r = get_process_comm(pid, &t2);
+                        if (r < 0)
+                                return r;
+
+                        mfree(t);
+                        t = strjoin("[", t2, "]");
+                        if (!t)
+                                return -ENOMEM;
+                }
+
+                bool eight_bit = (flags & PROCESS_CMDLINE_USE_LOCALE) && !is_locale_utf8();
+
+                ans = escape_non_printable_full(t, max_columns, eight_bit);
+                if (!ans)
                         return -ENOMEM;
-        }
 
-        delete_trailing_chars(t, WHITESPACE);
+                (void) str_realloc(&ans);
 
-        bool eight_bit = (flags & PROCESS_CMDLINE_USE_LOCALE) && !is_locale_utf8();
+        } else {
+                _cleanup_strv_free_ char **args = NULL;
 
-        ans = escape_non_printable_full(t, max_columns, eight_bit);
-        if (!ans)
-                return -ENOMEM;
+                assert(!(flags & PROCESS_CMDLINE_USE_LOCALE));
+
+                for (;;) {
+                        _cleanup_free_ char *arg = NULL, *escaped = NULL;
+
+                        r = read_nul_string(f, max_length, &arg);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                break;
+
+                        escaped = shell_maybe_quote(arg, ESCAPE_EMPTY);
+                        if (!escaped)
+                                return -ENOMEM;
+
+                        r = strv_consume(&args, escaped ? TAKE_PTR(escaped) : TAKE_PTR(arg));
+                        if (r < 0)
+                                return r;
+                }
+
+                if (args)
+                        ans = strv_join(args, " ");
+                else {
+                        if (!(flags & PROCESS_CMDLINE_COMM_FALLBACK))
+                                return -ENOENT;
+
+                        /* Kernel threads have no argv[] */
+                        _cleanup_free_ char *arg = NULL, *arg2 = NULL;
+
+                        r = get_process_comm(pid, &arg);
+                        if (r < 0)
+                                return r;
+
+                        arg2 = strjoin("[", arg, "]");
+                        if (!arg2)
+                                return -ENOMEM;
+
+                        ans = shell_maybe_quote(arg2, 0);
```

Comment:
we almost always prefer C style escaping over shell encoding. Shouldn't we continue to stick to that, and only use shell style escaping when we really know we want it?

or in other words: make the escaping style configured via a flag? I mean, there's already a flags param, might as well extend that?